File: Python/_0847_Shortest_Path_Visiting_All_Nodes/Solution.py
# coding=utf-8
# Date: 2021/8/6 10:34
from typing import List


# 执行用时：648 ms, 在所有 Python3 提交中击败了16.86%的用户内存消耗：16.2 MB, 在所有 Python3 提交中击败了70.93%的用户
class Solution:
    def shortestPathLength(self, graph: List[List[int]]) -> int:
        n, INF = len(graph), int(1e8)
        f = [[INF for _ in range(n)] for _ in range(1 << n)]

        q = []
        for i in range(n):
            f[1 << i][i] = 0
            q.append([1 << i, i])

        while len(q) != 0:
            t = q[0]
            q.pop(0)

            x, y = t[0], t[1]
            for z in graph[y]:
                s = x | 1 << z
                if f[s][z] > f[x][y] + 1:
                    f[s][z] = f[x][y] + 1
                    q.append([s, z])

        res = INF
        for i in range(n):
            res = min(res, f[(1 << n) - 1][i])
        return res


if __name__ == "__main__":
    print(Solution().shortestPathLength([[1, 2, 3], [0], [0], [0]]))  # 4

File: Java/src/_0847_Shortest_Path_Visiting_All_Nodes/Solution.java
package _0847_Shortest_Path_Visiting_All_Nodes;

import java.util.Arrays;
import java.util.LinkedList;
import java.util.Queue;

/**
 * Date: 2021/8/6 10:22
 * Content:
 * 执行用时：9 ms, 在所有 Java 提交中击败了78.31%的用户
 * 内存消耗：38.6 MB, 在所有 Java 提交中击败了39.16%的用户
 */
public class Solution {

    final static int INF = (int) 1e8;

    public int shortestPathLength(int[][] graph) {
        int n = graph.length;
        int[][] f = new int[1 << n][n];
        for (int i = 0; i < 1 << n; i++) Arrays.fill(f[i], INF);

        Queue<int[]> q = new LinkedList<>();
        for (int i = 0; i < n; i++) {
            f[1 << i][i] = 0;
            q.add(new int[]{1 << i, i});
        }

        while (!q.isEmpty()) {
            int[] t = q.remove();

            for (int z : graph[t[1]]) {
                int s = t[0] | 1 << z;
                if (f[s][z] > f[t[0]][t[1]] + 1) {
                    f[s][z] = f[t[0]][t[1]] + 1;
                    q.add(new int[]{s, z});
                }
            }
        }

        int res = INF;
        for (int i = 0; i < n; i++)
            res = Math.min(res, f[(1 << n) - 1][i]);
        return res;
    }

    public static void main(String[] args) {

        int[][] graph = {
                {1, 2, 3},
                {0},
                {0},
                {0},
        };
        System.out.println((new Solution()).shortestPathLength(graph));  // 4
    }
}
